Refactor subpath out of pkg/util/mount

codenrhoden opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
 /kind cleanup


**What this PR does / why we need it**:
This patch moves subpath functionality out of pkg/util/mount and into a
new package pkg/util/subpath. NSEnter funtionality is maintained.

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #72290

**Special notes for your reviewer**:
This PR has multiple commits in order to preserve the git history of files that were copied (and renamed) from `pkg/util/mount` to `pkg/volume/util/subpath`. In order to view changes that were done as part of that move, be sure to look at only the final commit in the patchset, otherwise they will look like completely new files.

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

codenrhoden commented:
/kind cleanup

codenrhoden commented:
/assign @msau42

codenrhoden commented:
/test pull-kubernetes-kubemark-e2e-gce-big

msau42 commented:
/assign @jsafrane @andyzhangx 

Thanks for working on this! This is looking good to me.

Is it possible to put this package under pkg/volume/util/subpath?  I'm not sure if you may hit import cycle issues, if so, then it's fine where it is.

msau42 commented on pkg/util/mount/exec_mount.go:
```diff
@@ -144,14 +144,6 @@ func (m *execMounter) EvalHostSymlinks(pathname string) (string, error) {
-	return m.wrappedMounter.CleanSubPaths(podDir, volumeName)
-}
-
 func (m *execMounter) SafeMakeDir(pathname string, base string, perm os.FileMode) error {
```
@jsafrane we added SafeMakeDir as part of the subpath work.  Should that go inside the subpath package even though it's more of a file util?

jsafrane replied:
+1. And take `DoSafeOpen`, `LockAndCheckSubPathWithoutSymlink` and similar functions with it, if possible. They're not related to mounter (neither subpaths, but that's the only reason they exist).

codenrhoden replied:
👍

codenrhoden commented:
@msau42 Thanks! I'll see if I can move this package as you asked.

I'm hitting one testing issue only seen in pull-kubernetes-kubemark-e2e-gce-big where https://github.com/kubernetes/kubernetes/pull/74393/files#diff-450e811a4953f760ff1594ede8b2037eR758 is panic'ing because `subpather` is `nil`. But only in one single test. Still trying to sort that one out. =)

jsafrane commented on cmd/kubelet/app/server.go:
```diff
@@ -371,6 +373,8 @@ func UnsecuredDependencies(s *options.KubeletServer) (*kubelet.Dependencies, err
 		}
 		mounter = mount.NewNsenterMounter(s.RootDirectory, ne)
+		// NSenter only valid on Linux
+		subpather = subpath.NewNSEnter()
```
nit: I'd prefer if there was `subpath.NewNSEnter(ne)`, so you don't need to export Ne from the mounter.

codenrhoden replied:
Make sense. I think it's a little weird to pass the `ne` object that will also be embedded in the `mounter` object that I am now going to pass as well, but it definitely keeps us from having to export it. will do.

jsafrane commented on cmd/kubelet/app/server.go:
```diff
@@ -363,6 +364,7 @@ func UnsecuredDependencies(s *options.KubeletServer) (*kubelet.Dependencies, err
 	}
 
 	mounter := mount.New(s.ExperimentalMounterPath)
+	subpather := subpath.New()
```
Not sure if it's possible, but can you pass mounter to subpather here (`subpath.New(mounter)`) instead of passing it in every `PrepareSafeSubpath` / `CleanSubPaths`?

codenrhoden replied:
I can. Good idea.

jsafrane commented on pkg/util/subpath/subpath.go:
```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2014 The Kubernetes Authors.
```
Wrong year (applies to all new files in this PR?)

codenrhoden replied:
Yeah, I wasn't exactly sure for this, as all the code is cut and paste from `/pkg/util/mount`. In that case, thought maybe it was better to keep copyright for the oldest code in the file. So I kept the year from whatever Linux/Windows file I was bringing the code over from. I'm still not sure what the right policy for that is.

codenrhoden replied:
Just as a final clarification, now that most of the files that are in `pkg/volume/util/subpath` are originally copied from `pkg/util/mount` and preserve git history, I only set the Copyright date on the actually new file, which is `subpath.go`.

codenrhoden reviewed (commented):
I'll work on these changes today. I'm still trying to track down the one failing test. I'm still confused why it happens, because it seems like `GetSubpather()` is returning `nil` in a specific case when it shouldn't.

What's weird is that `getSubpather` is implemented in 4 different controllers, and in 3 of those contorollers it does return `nil`. **But** so does `GetMounter()`, and the subpather call where the object is `nil` is replacing a former use of `GetMounter()`, so the behavior should have been the same.

I haven't been able to recreate locally, still working on that. Main difference I've seen in tests I can run with my KIND cluster and the `gce-big` test is that it does use the experimintal mounter path flag. I'm wondering if that has anything to do with it, but haven't nailed it down yet.

codenrhoden commented:
/test pull-kubernetes-e2e-kind

codenrhoden commented:
I finally found the test issue -- had never had to look into the `kubemark` code before, and found the issue in there. I think I've addressed all the comments now, except for the Copyright year issue. Just need some clarification on whether all new files should say 2019, or if they should maintain the oldest year from which the code came from.

I do expect tests to pass now.

codenrhoden commented:
I got some clarification from @dims, and I have no updated the copyright on all new files to 2019.

codenrhoden commented:
/priority important-soon

msau42 commented:
Is it possible to retain the commit history of the important files that were split (ie mount_linux/nsenter/windows/test)?  Maybe try this out in another branch just in case, but I think something like this should work:
1. git mv pkg/util/mount/mount_linux.go pkg/volume/util/subpath/mount_linux.go
2. cp pkg/volume/util/subpath/mount_linux.go pkg/volume/util/subpath/subpath_linux.go
3. repeat steps 1-2 for all the necessary files
4. git add pkg/volume/util/subpath
5. git commit -m "copy mount files to subpath util directory"
6. git mv pkg/volume/util/subpath/mount_linux.go pkg/util/mount/mount_linux.go.  repeat for all the files
7. git commit -m "move original mount files back"
8. apply edits on top of subpath_linux.go (copy the files in this pr)

codenrhoden commented:
@msau42 It's possible to do something like that. I've looked into it before, but never done it.

The steps you outlined for me didn't work, but following steps [here](https://stackoverflow.com/questions/16937359/git-copy-file-preserving-history) do. It would turn this patch into a series of 5 commits, with all but the final commit being "broken" from a build perspective. I cannot make it a simpler series of commits (by rebasing, I tried), as it requires creating two parallel histories (two commits) then merging them together (1 commit) a final rename (1 commit) and then applying my changes on top of that (1 commit).

The 4 immediate commits would look really confusing, and tests would not pass on them, but the 5th commit would look like exactly what we want on it's own, showing the changes to each file, and with things like `git blame` showing all the original history.

I can give this a shot if you like. I'd probably push a parallel branch for you to look at before overwriting the branch for this PR. I'd just be worried to go through the effort but then find we can't take a PR with 5 commits and a weird history like that. I personally think the effort is worth it, but it's more a project guideline thing to me. Thoughts?

msau42 commented:
The subpath code is substantial with many bug fixes so I think it is worth it.  Feel free to push a parallel branch first to test things out.  (btw I am on git 2.21.0 and those steps seemed to have worked for me)

codenrhoden commented:
@msau42 So [here](https://github.com/kubernetes/kubernetes/compare/master...codenrhoden:refactor-subpath-githistory) is what the new PR would look like, with preserving git history. It has 5 commits. Unfortunately, if you just use GitHub's review diffs, it picks up all the `pkg/volume/util/subpath/*` files as new files, and doesn't show you what has changed. I suppose this makes sense, as they are actually new files. `git blame` on them shows all the history, though. However, you can also review the commits separately, and if you do look at the last commit in the patchset, you **will** see all the changes that were made, and it makes it way easier to review and have confidence that nothing was missed or changed where we didn't mean to. Let me know if I should convert my PR over to this version.

codenrhoden commented:
> The subpath code is substantial with many bug fixes so I think it is worth it. Feel free to push a parallel branch first to test things out. (btw I am on git 2.21.0 and those steps seemed to have worked for me)

Hmm, I missed this comment while I was working on this other branch... I'm on git 2.17.1. I'll upgrade and see what happens. Would be nice to not have a bunch of extra commits!

msau42 commented:
> Would be nice to not have a bunch of extra commits!

Just to clarify, the sequence I suggested still has two extra commits (to move the files to the new location and back)

I took a look at your branch and it looks good to me! I just had one comment on a recent commit that got lost in the rebase

codenrhoden committed:
Copy mount files to pkg/volume/util/subpath

Files in pkg/util/mount that contain significant code implementation for
subpaths are moved to a new package at pkg/volume/util/subpath. This
move is done in order to preserve git history.

codenrhoden committed:
Move original mount files back

Move original mount files back into pkg/util/mount. This move is done to
preserve git history.

codenrhoden committed:
Refactor subpath out of pkg/util/mount

This patch moves subpath functionality out of pkg/util/mount and into a
new package pkg/volume/util/subpath. NSEnter funtionality is maintained.

codenrhoden commented:
Git history is now preserved, and all tests are passing. Ready for re-review, @msau42, @jsafrane .

We'll have to get some of the other required people from OWNERS files to take a look, which might be a challenge. This touches a lot of different areas. Since it does, merge conflicts are likely to pop up often!

msau42 commented on pkg/volume/util/subpath/subpath_linux.go:
```diff
@@ -63,681 +44,44 @@ const (
-		}
-	}
-	return mountErr
+type subpath struct {
```
can you also add this line to each implementation so we can statically check that the interface is implemented?
```
var _ Interface = &subpath{}
```

codenrhoden replied:
That's what https://github.com/kubernetes/kubernetes/pull/74393/commits/1da806942d23b0f7a2b351bfe60584f4f9f717ac#diff-8108985484b2fcf3c0170c3e6d32d97fR72 does, no? Just like the original mount code, it only needs to be in `subpath.go`, not each OS specific file, as it will only get picked up when `GOOS` is set correctly.

msau42 replied:
Ah ok, I think nsenter still needs one though since it's also linux.

codenrhoden replied:
Good catch. It does. Will update once I see results of `pull-kubernetes-e2e-gce-csi-serial`

msau42 commented:
@jsafrane can you help test containerized kubelet?

msau42 commented:
/test pull-kubernetes-e2e-gce-csi-serial

codenrhoden commented:
/test pull-kubernetes-integration

jsafrane commented:
Tested with containerized kubelet, no regressions.

msau42 commented:
/lgtm
/approve

Thanks so much for taking this on!

msau42 commented:
/assign @derekwaynecarr @jsafrane @shyamjvs 
for approval

derekwaynecarr commented:
@jsafrane thank you for confirming https://github.com/kubernetes/kubernetes/pull/74393#issuecomment-467929709

The kubelet changes lgtm as well.

/approve

shyamjvs commented:
/approve
(for the kubemark part)

saad-ali reviewed (approved):
Looks great. Thank you for this clean up!!

/approve
/lgtm
